File: [1]Tutorial_for_Python_and_R/7_Decision_Tree_Regression.py
# -*- coding: utf-8 -*-

# Decision Tree Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

import os

# Set the directory
currentPath = os.getcwd()
print(currentPath)
os.chdir('C:\\Users\\KYY\\Desktop\\ML_Ato_Z\\[0]Data')


# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
# make X -> matrix, y -> vector
X = dataset.iloc[:, 1:2].values   # don't need 'Position' column  # vector -> matrix
y = dataset.iloc[:, -1].values


# Splitting the dataset into the Training set and Test set
"""
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2,
                                                    random_state = 0)
"""

# Feature Scaling
# SVR library we use doesn't do feature scaling
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_Y = StandardScaler()
X= sc_X.fit_transform(X)
y = sc_Y.fit_transform(y)
"""


# Fitting the Decision Tree Regression to the dataset
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(criterion="mse", random_state = 0)   # default criterion is 'mse', 
regressor.fit(X, y)

# Predicting a new result with Decision Tree Regression
y_pred = regressor.predict(6.5)
# [6.5] = vector, not array.  [[6.5]]: an array of one line in one column


# Visualizing the SVR results (for higher resolution and smotther curve)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)

plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title('Truth or Bluff (Decision Tree Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

File: [1]Tutorial_for_Python_and_R/7_Decision_Tree_Regression.R
# Decision Tree Regression

# Set the directory
setwd('C:/Users/KYY/Desktop/ML_Ato_Z/[0]Data')

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]


# Feature Scaling - included in Linear regression package
# training_set = scale(training_set)
# test_set = scale(test_set)


# Fitting Decision Tree Regression to the dataset
install.packages('rpart')
library(rpart)
# press F1 for short cut to 'help'
regressor = rpart(formula = Salary ~ ., 
                data = dataset,
                control = rpart.control(minsplit = 3))


# Predicting a new result with Decision Tree Regression
y_pred = predict(regressor, newdata = data.frame(Level = 6.5))
y_pred



# Visualizing the Decision Tree Regression results (for higher resolution and smoother curve)
install.packages('ggplot2')
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)

ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = x_grid, y = predict(regressor, newdata = data.frame(Level = x_grid))),
            colour = 'blue') + 
  ggtitle('Truth or Bluff (Decision Tree Regression)') + 
  xlab('Level') + 
  ylab('Salary')